### ISSUE TITLE
Union of variant setters uses getter type for assignment

### ISSUE THREAD

**Issue by thw0rted** [microsoft/TypeScript#45376]
# Bug Report

### 🔎 Search Terms

variant accessor setter union

### 🕗 Version & Regression Information

- I was unable to test this on prior versions because: variant accessors only landed in 4.3

### ⏯ Playground Link

[Example from below](https://www.typescriptlang.org/play?#code/MYGwhgzhAEDyB2BTaBvAUNaBzRAXaADgE4D2BAFAJQBc0EuRAlvFqtEXgK5HzQBEfANzQAvhjp5CpChFr0mLaAB9o8TgFsARoiKU2YsWlCQYAFQDuJVOJz5iZKnIbNWKdlx78ho8REn2ZJwVWFTUtHT03AzQ0ABNEYw5oYBJ4emhOAEZaBEQlCxJBNCyAOgDoAF5oABYAJiKgA)

### 💻 Code

```ts
class One {
  get prop(): string { return ""; }
  set prop(s: string | number) { }
}

class Two {
  get prop(): string { return ""; }
  set prop(s: string | number) { }
}

declare const u1: One|Two;
u1.prop = 42; //Type 'number' is not assignable to type 'string'.(2322)
```

I didn't include it in the playground, but behavior is the same for `class` or `interface`.

### 🙁 Actual behavior

Assignment of number is flagged as an error, even though both sides of the union allow assignment of a number.

### 🙂 Expected behavior

Assignment should not be flagged.

----

Originally [reported by @ws93](https://github.com/microsoft/TypeScript/pull/42425#issuecomment-894715295)
